website_and_docs/content/documentation/grid/advanced_features/external_datastore_sessions.en.md

```diff
@@ -0,0 +1,272 @@
+---
+title: "External datastore for sessions"
+linkTitle: "External datastore for sessions"
+weight: 5
+---
+
+## Table of Contents
+* [Why an external datastore](#why-an-external-datastore)
+* [The solution](#the-solution)
+* [Setup](#setup)
+* [Database backed Session Map](#database-backed-session-map) 
+    * [Steps](#steps)
+* [Redis backed Session Map](#redis-backed-session-map) 
+    * [Steps](#steps)
+
+## Why an external datastore ?
+
+When you are running a distributed grid that comprises of:
+
+* Event Bus
+* New Session Queue
+* Session Map
+* Distributor
+* Router
+* Node
+
+you may want to scale these individual components horizontally (i.e., run more than one instances of the same component. For e.g., two `Session Map`) to cater to your grid's increased load.
+
+What would happen if we run two instances of a `Session Map`? 
+
+By default, a `Session Map` component uses an in-memory data store to store sessions information (which session is running on which node).
+
+So if there are two `Session Map` instances, then we end up with 2 in-memory data store instances which are not shared.
+
+This can cause a lot of issues such as:
+
+* Sessions data not being shared by both the instances
+* Session data query goes to the wrong instance of `Session Map`
+* Sessions data is lost if one of the instances goes down
+
+In short we now have a consistency problem that is so closely associated with the notion of microservices.
```

Comment:
With the current Grid 4, this is not accurate because it is not possible to run two `SessionMap` instances. Well, it is technically possible, but the `Router` can only talk to only one `SessionMap` instance.

The reasoning behind backing the data from the `SessionMap` is to allow the user to take down the `SessionMap` instance and bring a new one without loosing the current data.

As you can see, the topic of a high availability Grid has lots of room for improvement.

Reply:
Wouldn't it work fine if we put up a reverse proxy/load balancer (maybe something like a haproxy or nginx) which just does routing on a round robin basis? In that case, the router would just connect to the load balancer that fronts the session map and so it should be possible no ?

Reply:
Yes, but that is out of the scope of this doc for now.

That is what I tried to say with
>As you can see, the topic of a high availability Grid has lots of room for improvement.

We need to figure out ways for users to have a high availability Grid. What you propose is an alternative.

Reply:
> We need to figure out ways for users to have a high availability Grid. 

I feel that if Grid helps users to leverage High Availability Tools (which are meant for that purpose) then we wouldn't need to build them (I hope that we aren't looking at that) and so the ask of high availability would have been resolved right ?

Reply:
Yes, we should reuse existing tools as much as possible. But as the current state, we can only document what we know that works. 

A future step would be to figure out what tools are needed and how to combine them with Grid to make it highly available. 

We can continue this conversation in the #selenium-tlc channel on Slack 🙏

Reply:
Fixed